### ISSUE TITLE
Repeated recompilation in jax.grad / jax.value_and_grad with dft.RKS causes severe wall-time blowup.

### ISSUE THREAD

**Issue by wsr1999** [fishjojo/pyscfad#99]
# Repeated `jax.grad` with `dft.RKS` causes wall time blowup; `scf.RHF` unaffected

## Summary

Repeated calls to `jax.grad` (or `jax.value_and_grad`) with `dft.RKS` trigger severe recompilation overhead and wall-time blowup. The same pattern is not observed with `scf.RHF`. I reproduced this both in a minimal repeated-gradient test and in a practical geometry optimization workflow. The issue appears to be specific to the DFT/XC AD path rather than SCF convergence itself.

## Environment

pyscfad 0.2.0, pyscf 2.12.1, jax 0.8.3

## MRE 1: Repeated `jax.grad`

```python
from pyscfad import gto, dft, scf, config
config.update("pyscfad_scf_implicit_diff", True)
import jax, time

mol = gto.Mole()
mol.atom = 'C 0 0 -0.5265; O 0 0 0.6644; H 0 0.9369 -1.1129; H 0 -0.9369 -1.1129'
mol.unit = 'Angstrom'; mol.basis = 'def2-svp'; mol.cart = True
mol.build(trace_exp=False, trace_ctr_coeff=False)

def e_rhf(mol):
    mf = scf.RHF(mol); mf.kernel(); return mf.e_tot

def e_pbe(mol):
    mf = dft.RKS(mol); mf.xc = 'pbe'; mf.kernel(); return mf.e_tot

for name, fn in [("RHF", e_rhf), ("PBE", e_pbe)]:
    print(f"\n=== {name} ===")
    for i in range(5):
        t0 = time.time()
        g = jax.grad(fn)(mol)
        print(f"  call {i}: {time.time()-t0:.1f}s")
```

```
=== RHF ===
  call 0: 5.3s    (JIT compilation)
  call 1: 0.6s
  call 2: 0.6s
  call 3: 0.7s
  call 4: 0.7s

=== PBE ===
  call 0: 50.7s
  call 1: 151.1s
  call 2: 186.2s
  call 3: 187.0s
  call 4: 189.4s
```

## MRE 2: Practical impact — geometry optimization

H₂O/def2-SVP geometry optimization using PyBerny. RHF completes in 6s; PBE takes 85s with the last step spiking to 71.5s.

```python
import warnings
warnings.simplefilter("ignore")
import jax, time
from pyscfad import gto, dft, scf, config
config.update("pyscfad_scf_implicit_diff", True)
from pyscf.geomopt.berny_solver import optimize, to_berny_geom
from berny import Berny

def energy_rhf(mol):
    mf = scf.RHF(mol); mf.kernel(); return mf.e_tot

def energy_pbe(mol):
    mf = dft.RKS(mol); mf.xc = 'pbe'; mf.kernel(); return mf.e_tot

def run(energy_fn, name, atom, basis):
    val_and_grad = jax.value_and_grad(energy_fn)
    def solver(geom):
        mol = gto.Mole(); mol.verbose = 0; mol.atom = geom; mol.basis = basis
        mol.build(trace_exp=False, trace_ctr_coeff=False)
        e, jac = val_and_grad(mol)
        return e, jac.coords
    mol0 = gto.Mole(); mol0.build(atom=atom, basis=basis)
    optimizer = Berny(to_berny_geom(mol0))
    t_start = time.time()
    for i, geom in enumerate(optimizer):
        t0 = time.time()
        e, g = solver(list(geom))
        optimizer.send((e, g))
        print(f"  {name} step {i+1}: {float(e):.10f}  ({time.time()-t0:.1f}s)")
    print(f"  {name} total: {time.time()-t_start:.1f}s\n")

atom = 'O 0 0 0.1173; H 0 0.7572 -0.4692; H 0 -0.7572 -0.4692'
run(energy_rhf, "RHF", atom, 'def2-svp')
run(energy_pbe, "PBE", atom, 'def2-svp')
```

```
  RHF step 1: -75.9609839871  (4.8s)
  RHF step 2: -75.9613237854  (0.6s)
  RHF step 3: -75.9613383979  (0.5s)
  RHF total: 6.0s

  PBE step 1: -76.2720009615  (5.1s)
  PBE step 2: -76.2725534615  (4.0s)
  PBE step 3: -76.2725822975  (4.1s)
  PBE step 4: -76.2725826615  (71.5s)
  PBE total: 84.7s
```

## Notes

- PBE0 (hybrid) shows the same behavior as PBE (pure GGA).
- Energies/gradients appear numerically reasonable; this is a performance/recompilation issue.
- The description of this issue was refined by generative AI.

**Comment by fishjojo** [comment]
@claude review and suggest fix.
